csrc/id_model/id_graph.cpp

```diff
@@ -0,0 +1,1034 @@
+  TORCH_INTERNAL_ASSERT(
+      disjoint_set_pair.second,
+      "\nId group could not be found in graph associated with: ",
+      id->toString(),
+      "\n");
+  return disjoint_set_pair.first;
+}
+
+ExprGroups IdGraph::toGroups(const VectorOfUniqueEntries<Expr*>& exprs) const {
+  ExprGroups expr_groups;
+  for (auto expr : exprs) {
+    expr_groups.pushBack(toGroup(expr));
+  }
+  return expr_groups;
+}
+
+IdGroups IdGraph::toGroups(
+    const VectorOfUniqueEntries<IterDomain*>& ids) const {
+  IdGroups id_groups;
+  for (auto id : ids) {
+    id_groups.pushBack(toGroup(id));
+  }
+  return id_groups;
+}
+
+std::vector<IdGroup> IdGraph::outputGroups(ExprGroup expr) const {
+  std::vector<IdGroup> output_groups;
+  for (auto id_output :
+       ir_utils::filterByType<IterDomain>(expr->front()->outputs())) {
+    output_groups.push_back(toGroup(id_output));
+  }
+  return output_groups;
+}
+
+std::vector<IdGroup> IdGraph::inputGroups(ExprGroup expr) const {
+  std::vector<IdGroup> input_groups;
+  for (auto id_input :
+       ir_utils::filterByType<IterDomain>(expr->front()->inputs())) {
+    input_groups.push_back(toGroup(id_input));
+  }
+  return input_groups;
+}
+
+ExprGroups IdGraph::allUsesOf(const IdGroups& of) const {
+  ExprGroups to_visit;
+  for (auto of_id_group : of) {
+    auto group_uses_pair = iterDomainGroupUses(of_id_group);
+    if (group_uses_pair.second) {
+      to_visit.pushBack(group_uses_pair.first);
+    }
+  }
+
+  ExprGroups visited;
+  while (to_visit.size() > 0) {
+    auto current_expr = to_visit.popFront();
+    visited.pushBack(current_expr);
+    auto output_ids = outputGroups(current_expr);
+    for (auto output_id : output_ids) {
+      auto group_uses_pair = iterDomainGroupUses(output_id);
+      if (!group_uses_pair.second) {
+        continue;
+      }
+      for (auto group_use : group_uses_pair.first) {
+        if (visited.has(group_use)) {
+          continue;
+        }
+        to_visit.pushBack(group_use);
+      }
+    }
+  }
+
+  return visited;
+}
+
+ExprGroups IdGraph::allDefinitionsOf(const IdGroups& of) const {
+  ExprGroups to_visit;
+  for (auto of_id_group : of) {
+    auto group_defs_pair = iterDomainGroupDefinitions(of_id_group);
+    if (group_defs_pair.second) {
+      to_visit.pushBack(group_defs_pair.first);
+    }
+  }
+
+  ExprGroups visited;
+  while (to_visit.size() > 0) {
+    auto current_expr = to_visit.popFront();
+    visited.pushBack(current_expr);
+    auto input_ids = inputGroups(current_expr);
+    for (auto input_id : input_ids) {
+      auto group_defs_pair = iterDomainGroupDefinitions(input_id);
+      if (!group_defs_pair.second) {
+        continue;
+      }
+      for (auto group_def : group_defs_pair.first) {
+        if (visited.has(group_def)) {
+          continue;
+        }
+        to_visit.pushBack(group_def);
+      }
+    }
+  }
+
+  return visited;
+}
+
+ExprGroups IdGraph::getExprsBetween(const IdGroups& from, const IdGroups& to)
```

Comment:
Just started reading the code, but would it make our life easier if we just define `IdGroups` as a subclass of `Val`, and `ExprGroups` as a subclass of `Expr`, and just use `StmtSort::getExprsBetween`?

Reply:
`StmtSort` requires SSA, the structure we're operating in isn't SSA.